x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/profile/TransportProfileHasPrivilegesAction.java

```diff
@@ -0,0 +1,125 @@
+        this.privilegeStore = privilegeStore;
+        this.profileService = profileService;
+        this.securityContext = securityContext;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    protected void doExecute(Task task, ProfileHasPrivilegesRequest request, ActionListener<ProfileHasPrivilegesResponse> listener) {
+        profileService.getProfileSubjects(request.profileUids(), ActionListener.wrap(profileSubjectsAndFailures -> {
+            if (profileSubjectsAndFailures.profileUidToSubject().isEmpty()) {
+                listener.onResponse(new ProfileHasPrivilegesResponse(Set.of(), Set.of()));
+                return;
+            }
+            final Set<String> hasPrivilegeProfiles = Collections.synchronizedSet(new HashSet<>());
+            final Set<String> errorProfiles = Collections.synchronizedSet(new HashSet<>(profileSubjectsAndFailures.failureProfileUids()));
+            final Runnable allDone = () -> listener.onResponse(new ProfileHasPrivilegesResponse(hasPrivilegeProfiles, errorProfiles));
+            final Collection<Map.Entry<String, Subject>> profileUidAndSubjects = profileSubjectsAndFailures.profileUidToSubject()
+                .entrySet();
+            final AtomicInteger counter = new AtomicInteger(profileUidAndSubjects.size());
+            assert counter.get() > 0;
+            resolveApplicationPrivileges(
+                request,
+                ActionListener.wrap(applicationPrivilegeDescriptors -> threadPool.generic().execute(() -> {
+                    for (Map.Entry<String, Subject> profileUidToSubject : profileUidAndSubjects) {
+                        final String profileUid = profileUidToSubject.getKey();
+                        final Subject subject = profileUidToSubject.getValue();
+                        authorizationService.checkPrivileges(
+                            subject,
+                            request.privilegesToCheck(),
+                            applicationPrivilegeDescriptors,
+                            ActionListener.wrap(privilegesCheckResult -> {
+                                if (privilegesCheckResult.allMatch()) {
+                                    hasPrivilegeProfiles.add(profileUid);
+                                }
+                                if (counter.decrementAndGet() == 0) {
+                                    allDone.run();
+                                }
+                            }, checkPrivilegesException -> {
+                                logger.debug(
+                                    new ParameterizedMessage("Failed to check privileges for profile [{}]", profileUid),
+                                    checkPrivilegesException
+                                );
+                                errorProfiles.add(profileUid);
+                                if (counter.decrementAndGet() == 0) {
+                                    allDone.run();
+                                }
+                            })
```

Comment:
The `counter` check is a good use case for `ActionListener#runAfter` which likely helps inline `allDone` as well.

Reply:
This is a nice suggestion. I have pushed: [577c22c](https://github.com/elastic/elasticsearch/pull/85898/commits/577c22c564053dde08bc6ec6e52deb837f6c5d6e) .